stan/math/prim/scal/meta/enable_if_generics.hpp

```diff
@@ -0,0 +1,810 @@
+template <typename Check>
+using disable_if = typename std::enable_if_t<!Check::value>;
+
+/**
+ * Template metaprogram for deducing if a method should be remove a
+ * a function from overload resolution.
+ *
+ * Returns a type void if all conditions are true and otherwise fails.
+ */
+template <class... Checks>
+using enable_if_all = std::enable_if_t<math::conjunction<Checks...>::value>;
+
+/**
+ * Template metaprogram for deducing if a method should be remove a
+ * a function from overload resolution.
+ *
+ * Returns a type void if any of the conditions are true and otherwise fails.
+ */
+template <class... Checks>
+using enable_if_any = std::enable_if_t<math::disjunction<Checks...>::value>;
+
+/**
+ * Template metaprogram for deducing if a method should be remove a
+ * a function from overload resolution.
+ *
+ * Returns a type void if all of the conditions are false.
+ */
+template <class... Checks>
+using disable_if_all = std::enable_if_t<!math::disjunction<Checks...>::value>;
+
+/**
+ * Template metaprogram for deducing if a method should be remove a
+ * a function from overload resolution.
+ *
+ * Returns a type void if any of the conditions are false.
+ */
+template <class... Checks>
+using disable_if_any = std::enable_if_t<!math::conjunction<Checks...>::value>;
+
+// Check whether the decayed types are the same
+template <typename T, typename S>
+struct is_same_decay
+    : std::integral_constant<
+          bool, std::is_same<std::decay_t<T>, std::decay_t<S>>::value> {};
+
+template <typename T, typename S>
+using enable_if_same = std::enable_if_t<is_same_decay<T, S>::value>;
+
+template <typename T, typename S>
+using disable_if_same = disable_if<is_same_decay<T, S>>;
+
+template <typename T, typename... Types>
+using enable_if_all_same = enable_if_all<is_same_decay<T, Types>...>;
+
+template <typename T, typename... Types>
+using disable_if_all_same = disable_if_all<is_same_decay<T, Types>...>;
+
+// Checks decayed (non-const/ref'd) type is arithmetic
+template <typename T>
+struct is_arithmetic_decay
+    : std::integral_constant<bool, std::is_arithmetic<std::decay_t<T>>::value> {
+};
+
+template <typename T>
+using enable_if_arithmetic = std::enable_if_t<is_arithmetic_decay<T>::value>;
+
+template <typename T>
+using disable_if_arithmetic = disable_if<is_arithmetic_decay<T>>;
+
+template <typename... Types>
+using enable_if_all_arithmetic = enable_if_all<is_arithmetic_decay<Types>...>;
+
+template <typename... Types>
+using enable_if_any_arithmetic = enable_if_any<is_arithmetic_decay<Types>...>;
+
+template <typename... Types>
+using disable_if_all_arithmetic = disable_if_all<is_arithmetic_decay<Types>...>;
+
+template <typename... Types>
+using disable_if_any_arithmetic = disable_if_any<is_arithmetic_decay<Types>...>;
+
+// Check if a type contains or is arithmetic
+template <typename T>
+struct is_contains_arithmetic
+    : std::integral_constant<
+          bool, std::is_arithmetic<scalar_type_t<std::decay_t<T>>>::value> {};
```

Comment:
All the structs here follow this structure where std::integral_constant is essentially `std::false_type` or `std::true_type`